test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,294 @@
+package docker_e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"celestiaorg/celestia-app/test/docker-e2e/dockerchain"
+
+	signaltypes "github.com/celestiaorg/celestia-app/v5/x/signal/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	tastoradockertypes "github.com/celestiaorg/tastora/framework/docker"
+	"github.com/celestiaorg/tastora/framework/testutil/wait"
+	tastoratypes "github.com/celestiaorg/tastora/framework/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+)
+
+// TestCelestiaAppBinarySwapUpgrade tests a simple upgrade from one image tag
+// to another by swapping the binary (usually used for minor version upgrade)
+func (s *CelestiaTestSuite) TestCelestiaAppBinarySwapUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app minor upgrade test in short mode")
+	}
+
+	tt := []struct {
+		Name           string
+		BaseImageTag   string
+		TargetImageTag string
+	}{
+		{
+			Name:           "v4.0.2-rc2 to v4.0.10",
+			BaseImageTag:   "v4.0.2-rc2",
+			TargetImageTag: "v4.0.10",
+		},
+		{
+			Name:           "v4.0.9-mocha to v4.0.10-mocha",
+			BaseImageTag:   "v4.0.9-mocha",
+			TargetImageTag: "v4.0.10-mocha",
+		},
+		{
+			Name:           "v4.0.9-arabica to v4.0.10-arabica",
+			BaseImageTag:   "v4.0.9-arabica",
+			TargetImageTag: "v4.0.10-arabica",
+		},
+	}
+
+	for _, tc := range tt {
+		s.Run(tc.Name, func() {
+			s.runBinarySwapUpgradeTest(tc.BaseImageTag, tc.TargetImageTag)
+		})
+	}
+}
+
+// TestCelestiaAppSignalDrivenUpgrade tests app version upgrade using the signaling mechanism.
+func (s *CelestiaTestSuite) TestCelestiaAppSignalDrivenUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app major upgrade test in short mode")
+	}
+
+	tag, err := dockerchain.GetCelestiaTagStrict()
+	s.Require().NoError(err)
+
+	tt := []struct {
+		Name             string
+		ImageTag         string
+		TargetAppVersion uint64
+	}{
+		{
+			Name:             "v2 to v3",
+			ImageTag:         tag,
+			TargetAppVersion: 3,
+		},
+		{
+			Name:             "v3 to v4",
+			ImageTag:         tag,
+			TargetAppVersion: 4,
+		},
+	}
+
+	for _, tc := range tt {
+		s.Run(tc.Name, func() {
+			s.runSignalDrivenUpgradeTest(tc.ImageTag, tc.TargetAppVersion)
+		})
+	}
+}
+
+// runBinarySwapUpgradeTest tests a binary swap (minor version) upgrade for celestia-app.
+// It initializes a chain with the given base image tag, verifies bank send functionality,
+// upgrades the chain to the target image tag, restarts it, and then verifies both the
+// new binary version and continued bank send functionality.
+func (s *CelestiaTestSuite) runBinarySwapUpgradeTest(BaseImageTag, TargetImageTag string) {
```

Comment:
this name is also confusing, because it sounds like we're not signalling, and just hardforking via switching binaries

we don't support that and shouldn't be doing that as its not expected to work

if this only upgrading a minor version, then we should just call it that

the issue I had with naming was that we just change this to a restart test as this is testing a scenario that will never happen IRL, where all nodes in the network upgrade to a minor version at the exact same time.

restarting is a useful test, since that's a scenario that does occur all the time and has occasionally been buggy. upgrading all nodes at the same time is not a useful test, and I still don't think we should do it.

Reply:
@coderabbitai are you activated here? if so create a followup issue out of this discussion to have an incompatibility test.

Reply:
a follwup is created: #5329